lib/Dialect/TritonNvidiaGPU/Transforms/FenceInsertion.cpp

```diff
@@ -50,21 +55,61 @@ struct FenceInsertionPass
                                .getEncoding()
                                .dyn_cast<ttg::MmaEncodingAttr>();
         auto isHopperEncoding = mmaEncoding && mmaEncoding.isHopper();
-        if (isHopperEncoding && (isa<ttg::ConvertLayoutOp>(a.getDefiningOp()) &&
-                                 ttg::isSharedEncoding(a)) ||
-            (isa<ttg::ConvertLayoutOp>(b.getDefiningOp()) &&
-             ttg::isSharedEncoding(b))) {
-
-          // TODO: check whether cluster fence is needed
-          OpBuilder builder(op);
+        if (isHopperEncoding && (canReachGeneric(a) || canReachGeneric(b))) {
           builder.create<ttng::FenceAsyncSharedOp>(op->getLoc(),
                                                    false /*bCluster*/);
         }
       }
     });
   }
-};
 
+private:
+  bool canReachGeneric(Value operand) {
+    static DenseSet<std::pair<Operation *, unsigned>> trace;
+    auto op = operand.getDefiningOp();
+    // reach convertlayout
+    if (op && isa<ttg::ConvertLayoutOp>(op) && ttg::isSharedEncoding(operand))
+      return true;
+    // root and not BlockArgument
+    if (!op && !isa<BlockArgument>(operand))
+      return false;
+    // op and not BlockArgument
+    if (op && !isa<BlockArgument>(operand)) {
+      for (auto v : op->getOperands()) {
+        if (canReachGeneric(v))
+          return true;
+      }
+    }
+    // reach BlockArgument
+    if (BlockArgument arg = dyn_cast<BlockArgument>(operand)) {
+      unsigned argNum = arg.getArgNumber();
+      Operation *argOwner = arg.getOwner()->getParentOp();
+      // suport ForOp
+      if (auto forOp = dyn_cast<scf::ForOp>(argOwner)) {
```

Comment:
I think it's not limited to a `forOp`, should be applied to any ops with `BlockArgument` right? It's OK to leave it with forOp though since we only pipeline  `forOp`

Reply:
Do we need to at least assert if argOwner is not ForOp?